main/src/com/google/refine/expr/functions/arrays/Zip.java

```diff
@@ -0,0 +1,134 @@
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+*/
+
+package com.google.refine.expr.functions.arrays;
+
+import static java.util.stream.Collectors.toCollection;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Properties;
+import java.util.stream.Collectors;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.node.ArrayNode;
+import com.google.common.collect.Streams;
+
+import com.google.refine.expr.EvalError;
+import com.google.refine.expr.ExpressionUtils;
+import com.google.refine.grel.ControlFunctionRegistry;
+import com.google.refine.grel.EvalErrorMessage;
+import com.google.refine.grel.Function;
+import com.google.refine.grel.FunctionDescription;
+
+public class Zip implements Function {
+
+    @Override
+    public Object call(Properties bindings, Object[] args) {
+        if (args.length >= 2) {
+            List<List> input = new ArrayList<>();
+            List<List> output;
+
+            for (Object v : args) {
+                List ziparg = new ArrayList<>();
+                if (v == null || !(v.getClass().isArray() || v instanceof List<?> || v instanceof ArrayNode)) {
+                    return new EvalError(
+                            EvalErrorMessage.expects_at_least_two_or_more_array_args(ControlFunctionRegistry.getFunctionName(this)));
+                }
+
+                if (v.getClass().isArray()) {
+                    ziparg.addAll(Arrays.asList((Object[]) v));
+                } else if (v instanceof ArrayNode) {
+                    ArrayNode a = (ArrayNode) v;
+                    if (a.isArray()) {
+                        for (JsonNode objNode : a) {
+                            ziparg.add(objNode);
+                        }
+                    }
+                } else {
+                    for (Object z : ExpressionUtils.toObjectList(v)) {
+                        ziparg.add(z);
+                    }
+                }
+                input.add(ziparg);
+            }
+
+            output = (List<List>) Streams.zip(input.get(0).stream(),
+                    input.get(1).stream(),
+                    (a, b) -> collectElements(a, b))
+                    .collect(Collectors.toList());
+
+            if (input.size() > 2) {
+                for (int i = 2; i < input.size(); i++) {
+                    output = (List<List>) Streams.zip(output.stream(),
+                            input.get(i).stream(),
+                            (a, b) -> collectElements(a, b))
+                            .collect(Collectors.toList());
+                }
+            }
+            return output;
```

Comment:
Then you can change the order of iteration so that you first iterate by element and then by input argument similar to the following:

```suggestion
            List<List> result = new ArrayList<>();
            boolean done = false;
            while (!done) {
                List element = new ArrayList<>();
                
                // Construct an output element by getting 
                // input element from each iterator
                // if any iterator is empty, we are done

                if (!done) {
                    result.add(element);
                }
            }
           return result;
```

Does that make sense? Let me know if you have any questions.

Reply:
@tfmorris Thanks for the feedback. Can you also confirm if the output of the zip is not limited to the shortest input. 
Is the below acceptable output of zip function?

-  zip([1,2], ['A','B','C'], ['X','Y','Z']) -> [[[1, A, X], [2, B, Y], [null, C, Z]]]

Reply:
Is there a reason that you'd prefer your alternative? If there's a user needs based reason, we should discuss it on the original issue where folks will see the discussion, but if it's just an implementation based preference, I think it would be better to make the implementation conform to the specification as currently given, i.e. the output for your example should be

`zip([1,2], ['A','B','C'], ['X','Y','Z']) -> [[[1, A, X], [2, B, Y]]]`

Reply:
@tfmorris The query came up as part of implementation. No change required in the zip working definition